File: frontend/src/lib/vacancy.ts
import type { VacancyPriority, VacancyStatus, WorkFormat } from '../types'

export const statusLabels: Record<VacancyStatus, string> = {
  interesting: 'Интересно',
  applied: 'Отклик',
  interview: 'Интервью',
  test: 'Тестовое',
  offer: 'Оффер',
  rejected: 'Отказ',
  archived: 'Архив',
}

export const priorityLabels: Record<VacancyPriority, string> = {
  low: 'Низкий',
  medium: 'Средний',
  high: 'Высокий',
}

export const workFormatLabels: Record<WorkFormat, string> = {
  remote: 'Удалённо',
  office: 'Офис',
  hybrid: 'Гибрид',
}

export const statusStyles: Record<VacancyStatus, string> = {
  interesting: 'bg-sky-50 text-sky-700 ring-sky-200',
  applied: 'bg-indigo-50 text-indigo-700 ring-indigo-200',
  interview: 'bg-amber-50 text-amber-800 ring-amber-200',
  test: 'bg-violet-50 text-violet-700 ring-violet-200',
  offer: 'bg-emerald-50 text-emerald-700 ring-emerald-200',
  rejected: 'bg-rose-50 text-rose-700 ring-rose-200',
  archived: 'bg-zinc-100 text-zinc-600 ring-zinc-200',
}

File: backend/app/schemas/vacancy.py
from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field, field_validator


class VacancySource(StrEnum):
    """Supported places where a vacancy can be found."""

    HH = "hh"
    LINKEDIN = "linkedin"
    TELEGRAM = "telegram"
    MANUAL = "manual"
    OTHER = "other"


class VacancyStatus(StrEnum):
    """Supported workflow statuses for a vacancy."""

    INTERESTING = "interesting"
    APPLIED = "applied"
    INTERVIEW = "interview"
    TEST = "test"
    OFFER = "offer"
    REJECTED = "rejected"
    ARCHIVED = "archived"


class VacancyPriority(StrEnum):
    """Supported priority levels for a vacancy."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class WorkFormat(StrEnum):
    """Supported work formats for a vacancy."""

    REMOTE = "remote"
    OFFICE = "office"
    HYBRID = "hybrid"


class VacancyFilters(BaseModel):
    """Optional query parameters for filtering the vacancy list."""

    search: str | None = Field(default=None, max_length=120)
    status: VacancyStatus | None = None
    priority: VacancyPriority | None = None
    work_format: WorkFormat | None = None
    source: VacancySource | None = None
    skill: str | None = Field(default=None, max_length=80)

    @field_validator("search", "skill")
    @classmethod
    def strip_non_empty_text(cls, value: str | None) -> str | None:
        """Trim optional text filters and reject blank values."""
        if value is None:
            return None

        stripped_value = value.strip()
        if not stripped_value:
            raise ValueError("value must not be blank")

        return stripped_value


class VacancyBase(BaseModel):
    """Shared vacancy fields used by create and read schemas."""

    company: str = Field(min_length=1, max_length=120)
    position: str = Field(min_length=1, max_length=160)
    url: str | None = Field(default=None, max_length=500)
    source: VacancySource = VacancySource.MANUAL
    status: VacancyStatus = VacancyStatus.INTERESTING
    priority: VacancyPriority = VacancyPriority.MEDIUM
    salary: str | None = Field(default=None, max_length=120)
    location: str | None = Field(default=None, max_length=120)
    work_format: WorkFormat = WorkFormat.REMOTE
    skills: list[str] = Field(default_factory=list)
    notes: str | None = None
    next_action: str | None = Field(default=None, max_length=240)


class VacancyCreate(VacancyBase):
    """Payload for creating a vacancy."""

    pass


class VacancyUpdate(BaseModel):
    """Payload for partially updating a vacancy."""

    company: str | None = Field(default=None, min_length=1, max_length=120)
    position: str | None = Field(default=None, min_length=1, max_length=160)
    url: str | None = Field(default=None, max_length=500)
    source: VacancySource | None = None
    status: VacancyStatus | None = None
    priority: VacancyPriority | None = None
    salary: str | None = Field(default=None, max_length=120)
    location: str | None = Field(default=None, max_length=120)
    work_format: WorkFormat | None = None
    skills: list[str] | None = None
    notes: str | None = None
    next_action: str | None = Field(default=None, max_length=240)


class VacancyRead(VacancyBase):
    """Vacancy representation returned by the API."""

    id: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
